{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metaphlan is a tool used for taxonomic profiling of metagenomic data. This notebook outlines the steps needed to convert metaphlan output into separate csv files compatible with R's phyloseq. https://github.com/biobakery/MetaPhlAn/wiki/MetaPhlAn-3.0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## To create a phyloseq object you need 3 tables:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Abundance table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| Otu | Sample_1 | Sample_2 |\n",
    "|-----|----------|----------|\n",
    "|Out1 |12461     | 463      |\n",
    "|Out2 |470       | 73       |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Taxonomy table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| Otu | Kingdom   | Phylum                | Class                 |  Order                | Family               | Genus               | Species                        |\n",
    "|-----|-----------|-----------------------|-----------------------|-----------------------|----------------------|---------------------|--------------------------------|\n",
    "|Out1 |k__Viruses |p__Viruses_unclassified|c__Viruses_unclassified|o__Viruses_unclassified|f__Polydnaviridae     |g__Bracovirus        |s__Cotesia_congregata_bracovirus|                             \n",
    "|Out2 |k__Archaea |p__Euryarchaeota       |c__Methanobacteria     |o__Methanobacteriales  |f__Methanobacteriaceae|g__Methanobrevibacter|s__Methanobrevibacter_smithii   |        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Sample table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| Otu     | Condition  | Treatment |\n",
    "|---------|------------|-----------|\n",
    "|Sample_1 |CD          | 1         |\n",
    "|Sample_2 |UC          | 2         |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sample metaphlan output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metaphlan gives us output in a compressed format by putting all the taxa identified in the ID column and the quantity of taxa present in each sample column. The best way to approach this dataset is to extract only the rows of the taxa you wish to look at. For example, if you want to look at the species level filter out all the rows which end in \"s__\" or for order filter the rows ending in \"o__\"."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|ID                                            |    Sample_1_profiled_metagenome | Sample_2_profiled_metagenome|\n",
    "|----------------------------------------------|---------------------------------|-----------------------------|\n",
    "|UNKNOWN                                                             |   92.02399| 91.02948                    |\n",
    "|k__Archaea                                                          |   0.052938| 0.045632                    |\n",
    "|k__Archaea\\p__Euryarhaeota                                          |   0.052938|0.045632                     |\n",
    "|k__Archaea\\p__Euryachaeota\\c__Methanobacteria                       |   0.052938|0.045632                     |\n",
    "|k__Archaea\\p__Euryarchaeota\\c__Methanobacteria\\o__Methanobacteriales|   0.052938|0.045632                     | "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_merged_table(df,merged_file_name):\n",
    "    '''Cleans the merged table by removing file extenstion suffix which gets added to sample columns during merge.'''\n",
    "    #df = pd.read_table(df,sep='\\t',engine='python')\n",
    "    df.columns = df.columns.str.replace(merged_file_name, '')\n",
    "    return df\n",
    "\n",
    "def get_taxa_columns(df,rank):\n",
    "    '''Splits ID into taxanomic ranks to make taxa table''' \n",
    "    df_taxa = df['ID'].str.split('|',expand=True)\n",
    "    taxa_cols = [\"Kingdom\",\"Phylum\",\"Class\",\"Order\",\"Family\",\"Genus\",\"Species\",\"Strain\"]\n",
    "    taxa_dict = {'Kingdom':1,\"Phylum\":2,\"Class\":3,\"Order\":4,\"Family\":5,\"Genus\":6,\"Species\":7,\"Strain\":8}\n",
    "    value = taxa_dict.get(rank)\n",
    "    taxa_cols=taxa_cols[0:value]\n",
    "    df_taxa.columns=taxa_cols\n",
    "    for col in df_taxa.columns:\n",
    "        df_taxa[col]=df_taxa[col].apply(trim_taxa_names)    \n",
    "    otu_index = []\n",
    "    for i in range(0, len(df)):\n",
    "        otu_index.append(\"Otu\"+str(i))\n",
    "    df_taxa['Otu']=otu_index \n",
    "    taxa_cols=[col for col in df_taxa.columns if 'Otu' not in col]\n",
    "    for col in taxa_cols:\n",
    "        df_taxa.at[df_taxa.index[-1], col] = 'Other'\n",
    "    return df_taxa\n",
    "\n",
    "def trim_taxa_names(x):\n",
    "    '''Removes leading characters before taxa ID e.g. s__ '''\n",
    "    match = re.sub(r'^[kpcofgs]__',\"\",str(x))\n",
    "    return match\n",
    "\n",
    "def get_sample_cols(df):\n",
    "    '''Finds and returns sample columns in dataframe, presuming sample names contain a number'''\n",
    "    r = re.compile(r'^.*[0-9].*$') #match column names that contain a number anywhere\n",
    "    sample_cols=[]\n",
    "    for col in df:\n",
    "        if(r.match(col)):\n",
    "            sample_cols.append(col)\n",
    "    return sample_cols\n",
    "\n",
    "def create_sample_df(abun_matrix):\n",
    "    sample_cols=get_sample_cols(abun_matrix)\n",
    "    sample_df=pd.DataFrame({'Sample':sample_cols})\n",
    "    sample_df['Behaviour'] = sample_df['Sample'].apply(get_behaviour)\n",
    "    # Add extra column to sample df so phyloseq ordination plots behave\n",
    "    sample_df['Type'] = 'murine'\n",
    "    return sample_df\n",
    "\n",
    "def add_otu_primary_key(df):\n",
    "    '''Adds otu primary key column to dataframe'''\n",
    "    otu_index = []\n",
    "    for i in range(0, len(df)):\n",
    "        otu_index.append(\"Otu\"+str(i))\n",
    "    df['Otu']=otu_index \n",
    "    return df\n",
    "\n",
    "def get_behaviour(x):\n",
    "    '''Creates sample_data columns based on sample name starting with letter'''\n",
    "    if x.startswith('E'):\n",
    "        return \"Extinction\"\n",
    "    elif x.startswith('R'):\n",
    "        return \"Resilient\"\n",
    "    elif x.startswith('S'):\n",
    "        return \"Susceptible\"\n",
    "    else:\n",
    "        return ''\n",
    "    \n",
    "def create_other_group(df,thresh):\n",
    "    '''From a given threshold or cut-off point it creates an \"Other\" row with the summed abundance of \n",
    "    species below the givedn threshold. Used to give idea of remainder of species excluded from plot legend'''\n",
    "    index_to_append = len(df)\n",
    "    df.at[index_to_append,'ID']= 'Other'\n",
    "    cols= df.columns \n",
    "    for col in cols:\n",
    "        if col!='ID':\n",
    "            df.loc[df['ID']=='Other', col]= np.sum(df.loc[(df[col]<=thresh) & (df['ID']!=\"other\")][col])\n",
    "    sample_cols = get_sample_cols(df)\n",
    "    ids=df['ID'].to_list()\n",
    "    df_thresh_removed = df[sample_cols].apply(lambda x: np.where(x <= thresh,0,x))\n",
    "    df_thresh_removed['ID']=ids\n",
    "    df=df_thresh_removed[(df_thresh_removed.iloc[:,0:-1] > 0).any(axis=1)]\n",
    "    df.reset_index(drop=True,inplace=True)\n",
    "    return df\n",
    "\n",
    "def create_sample_df(abun_matrix):\n",
    "    sample_cols=get_sample_cols(abun_matrix)\n",
    "    sample_df=pd.DataFrame({'Sample':sample_cols})\n",
    "    sample_df['Behaviour'] = sample_df['Sample'].apply(get_behaviour)\n",
    "    # Add extra column to sample df so phyloseq ordination plots behave\n",
    "    sample_df['Type'] = 'murine'\n",
    "    return sample_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Importing the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/linda/Projects/metaphlan'"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pool1_profile_known.txt  sample_df.csv\t\t      species_taxa.csv\r\n",
      "pool2_profile_known.txt  species_known_abundance.csv\r\n"
     ]
    }
   ],
   "source": [
    "!ls /home/linda/Projects/metaphlan/data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I have called these files profile \"known\" since the relative abundances are calculated out of the proportion of known reads that were classified. It is possible to estimate the relative abundances including the unknown fraction of reads by running metaphlan with the parameter --unknown-estimation. \n",
    "\n",
    "There are 2 files from 2 different sequencing pools which I am going to merge. These filed were created by metaphlan's merge_metaphlan_tables.py file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_k_1 = pd.read_table(\"data/pool1_profile_known.txt\", sep='\\t')\n",
    "df_k_2 = pd.read_table(\"data/pool2_profile_known.txt\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>E15_S89_known_profiled_metagenome</th>\n",
       "      <th>E19b_S88_known_profiled_metagenome</th>\n",
       "      <th>E28_S87_known_profiled_metagenome</th>\n",
       "      <th>E5b_S84_known_profiled_metagenome</th>\n",
       "      <th>E7_S78_known_profiled_metagenome</th>\n",
       "      <th>E8_S74_known_profiled_metagenome</th>\n",
       "      <th>R21b_S76_known_profiled_metagenome</th>\n",
       "      <th>R24_S82_known_profiled_metagenome</th>\n",
       "      <th>R5_S75_known_profiled_metagenome</th>\n",
       "      <th>R6_S81_known_profiled_metagenome</th>\n",
       "      <th>S11b_S73_known_profiled_metagenome</th>\n",
       "      <th>S11_S83_known_profiled_metagenome</th>\n",
       "      <th>S13_S79_known_profiled_metagenome</th>\n",
       "      <th>S22b_S90_known_profiled_metagenome</th>\n",
       "      <th>S22_S77_known_profiled_metagenome</th>\n",
       "      <th>S3b_S86_known_profiled_metagenome</th>\n",
       "      <th>S4_S85_known_profiled_metagenome</th>\n",
       "      <th>S9_S80_known_profiled_metagenome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>97.65801</td>\n",
       "      <td>99.37595</td>\n",
       "      <td>96.86904</td>\n",
       "      <td>93.72727</td>\n",
       "      <td>97.30751</td>\n",
       "      <td>96.97202</td>\n",
       "      <td>99.28079</td>\n",
       "      <td>99.51966</td>\n",
       "      <td>97.72520</td>\n",
       "      <td>99.77435</td>\n",
       "      <td>99.13651</td>\n",
       "      <td>97.13792</td>\n",
       "      <td>99.71291</td>\n",
       "      <td>99.81773</td>\n",
       "      <td>98.82742</td>\n",
       "      <td>96.79303</td>\n",
       "      <td>96.87159</td>\n",
       "      <td>98.24866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria</td>\n",
       "      <td>1.77270</td>\n",
       "      <td>35.18209</td>\n",
       "      <td>0.93580</td>\n",
       "      <td>1.73347</td>\n",
       "      <td>1.61261</td>\n",
       "      <td>1.83636</td>\n",
       "      <td>0.89332</td>\n",
       "      <td>1.38952</td>\n",
       "      <td>1.59392</td>\n",
       "      <td>2.32794</td>\n",
       "      <td>1.11673</td>\n",
       "      <td>2.29589</td>\n",
       "      <td>1.62908</td>\n",
       "      <td>21.00678</td>\n",
       "      <td>0.84331</td>\n",
       "      <td>11.85350</td>\n",
       "      <td>0.99145</td>\n",
       "      <td>2.12132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Actinobacteria</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>33.99356</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>20.50706</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>11.21542</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>33.99356</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>20.50706</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>11.21542</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>33.99356</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>20.50706</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>11.21542</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                           ID  \\\n",
       "0  k__Bacteria                                                                                  \n",
       "1  k__Bacteria|p__Actinobacteria                                                                \n",
       "2  k__Bacteria|p__Actinobacteria|c__Actinobacteria                                              \n",
       "3  k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales                         \n",
       "4  k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae   \n",
       "\n",
       "   E15_S89_known_profiled_metagenome  E19b_S88_known_profiled_metagenome  \\\n",
       "0  97.65801                           99.37595                             \n",
       "1  1.77270                            35.18209                             \n",
       "2  0.00000                            33.99356                             \n",
       "3  0.00000                            33.99356                             \n",
       "4  0.00000                            33.99356                             \n",
       "\n",
       "   E28_S87_known_profiled_metagenome  E5b_S84_known_profiled_metagenome  \\\n",
       "0  96.86904                           93.72727                            \n",
       "1  0.93580                            1.73347                             \n",
       "2  0.00000                            0.00000                             \n",
       "3  0.00000                            0.00000                             \n",
       "4  0.00000                            0.00000                             \n",
       "\n",
       "   E7_S78_known_profiled_metagenome  E8_S74_known_profiled_metagenome  \\\n",
       "0  97.30751                          96.97202                           \n",
       "1  1.61261                           1.83636                            \n",
       "2  0.00000                           0.00000                            \n",
       "3  0.00000                           0.00000                            \n",
       "4  0.00000                           0.00000                            \n",
       "\n",
       "   R21b_S76_known_profiled_metagenome  R24_S82_known_profiled_metagenome  \\\n",
       "0  99.28079                            99.51966                            \n",
       "1  0.89332                             1.38952                             \n",
       "2  0.00000                             0.00000                             \n",
       "3  0.00000                             0.00000                             \n",
       "4  0.00000                             0.00000                             \n",
       "\n",
       "   R5_S75_known_profiled_metagenome  R6_S81_known_profiled_metagenome  \\\n",
       "0  97.72520                          99.77435                           \n",
       "1  1.59392                           2.32794                            \n",
       "2  0.00000                           0.00000                            \n",
       "3  0.00000                           0.00000                            \n",
       "4  0.00000                           0.00000                            \n",
       "\n",
       "   S11b_S73_known_profiled_metagenome  S11_S83_known_profiled_metagenome  \\\n",
       "0  99.13651                            97.13792                            \n",
       "1  1.11673                             2.29589                             \n",
       "2  0.00000                             0.00000                             \n",
       "3  0.00000                             0.00000                             \n",
       "4  0.00000                             0.00000                             \n",
       "\n",
       "   S13_S79_known_profiled_metagenome  S22b_S90_known_profiled_metagenome  \\\n",
       "0  99.71291                           99.81773                             \n",
       "1  1.62908                            21.00678                             \n",
       "2  0.00000                            20.50706                             \n",
       "3  0.00000                            20.50706                             \n",
       "4  0.00000                            20.50706                             \n",
       "\n",
       "   S22_S77_known_profiled_metagenome  S3b_S86_known_profiled_metagenome  \\\n",
       "0  98.82742                           96.79303                            \n",
       "1  0.84331                            11.85350                            \n",
       "2  0.00000                            11.21542                            \n",
       "3  0.00000                            11.21542                            \n",
       "4  0.00000                            11.21542                            \n",
       "\n",
       "   S4_S85_known_profiled_metagenome  S9_S80_known_profiled_metagenome  \n",
       "0  96.87159                          98.24866                          \n",
       "1  0.99145                           2.12132                           \n",
       "2  0.00000                           0.00000                           \n",
       "3  0.00000                           0.00000                           \n",
       "4  0.00000                           0.00000                           "
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_k_1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lets look at genus level by using a regular expression to filter rows based on last occurrence of a letter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_k_1_genus=df_k_1[df_k_1['ID'].str.contains(r'\\|g__[^|]*$')]\n",
    "df_k_2_genus=df_k_2[df_k_2['ID'].str.contains(r'\\|g__[^|]*$')]\n",
    "df_k_1_genus.reset_index(drop=True,inplace=True)\n",
    "df_k_2_genus.reset_index(drop=True,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>E15_S89_known_profiled_metagenome</th>\n",
       "      <th>E19b_S88_known_profiled_metagenome</th>\n",
       "      <th>E28_S87_known_profiled_metagenome</th>\n",
       "      <th>E5b_S84_known_profiled_metagenome</th>\n",
       "      <th>E7_S78_known_profiled_metagenome</th>\n",
       "      <th>E8_S74_known_profiled_metagenome</th>\n",
       "      <th>R21b_S76_known_profiled_metagenome</th>\n",
       "      <th>R24_S82_known_profiled_metagenome</th>\n",
       "      <th>R5_S75_known_profiled_metagenome</th>\n",
       "      <th>R6_S81_known_profiled_metagenome</th>\n",
       "      <th>S11b_S73_known_profiled_metagenome</th>\n",
       "      <th>S11_S83_known_profiled_metagenome</th>\n",
       "      <th>S13_S79_known_profiled_metagenome</th>\n",
       "      <th>S22b_S90_known_profiled_metagenome</th>\n",
       "      <th>S22_S77_known_profiled_metagenome</th>\n",
       "      <th>S3b_S86_known_profiled_metagenome</th>\n",
       "      <th>S4_S85_known_profiled_metagenome</th>\n",
       "      <th>S9_S80_known_profiled_metagenome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae|g__Bifidobacterium</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>33.99356</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>20.50706</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>11.21542</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Eggerthellales|f__Eggerthellaceae|g__Adlercreutzia</td>\n",
       "      <td>0.01522</td>\n",
       "      <td>0.02269</td>\n",
       "      <td>0.00986</td>\n",
       "      <td>0.01337</td>\n",
       "      <td>0.02927</td>\n",
       "      <td>0.04169</td>\n",
       "      <td>0.01290</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.02503</td>\n",
       "      <td>0.05092</td>\n",
       "      <td>0.01569</td>\n",
       "      <td>0.03171</td>\n",
       "      <td>0.04299</td>\n",
       "      <td>0.01102</td>\n",
       "      <td>0.02144</td>\n",
       "      <td>0.01528</td>\n",
       "      <td>0.01784</td>\n",
       "      <td>0.03676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Eggerthellales|f__Eggerthellaceae|g__Asaccharobacter</td>\n",
       "      <td>0.06825</td>\n",
       "      <td>0.02827</td>\n",
       "      <td>0.02306</td>\n",
       "      <td>0.02648</td>\n",
       "      <td>0.02673</td>\n",
       "      <td>0.03658</td>\n",
       "      <td>0.02223</td>\n",
       "      <td>0.00543</td>\n",
       "      <td>0.03252</td>\n",
       "      <td>0.06713</td>\n",
       "      <td>0.02811</td>\n",
       "      <td>0.04405</td>\n",
       "      <td>0.03960</td>\n",
       "      <td>0.01511</td>\n",
       "      <td>0.00854</td>\n",
       "      <td>0.01129</td>\n",
       "      <td>0.01096</td>\n",
       "      <td>0.06209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Eggerthellales|f__Eggerthellaceae|g__Enterorhabdus</td>\n",
       "      <td>1.68923</td>\n",
       "      <td>1.13757</td>\n",
       "      <td>0.90287</td>\n",
       "      <td>1.69362</td>\n",
       "      <td>1.55661</td>\n",
       "      <td>1.75809</td>\n",
       "      <td>0.85820</td>\n",
       "      <td>1.38410</td>\n",
       "      <td>1.53637</td>\n",
       "      <td>2.20989</td>\n",
       "      <td>1.07292</td>\n",
       "      <td>2.22013</td>\n",
       "      <td>1.54649</td>\n",
       "      <td>0.47359</td>\n",
       "      <td>0.81333</td>\n",
       "      <td>0.61152</td>\n",
       "      <td>0.96265</td>\n",
       "      <td>2.02247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides</td>\n",
       "      <td>2.57402</td>\n",
       "      <td>1.03313</td>\n",
       "      <td>2.58441</td>\n",
       "      <td>5.52997</td>\n",
       "      <td>4.55413</td>\n",
       "      <td>2.52983</td>\n",
       "      <td>1.46428</td>\n",
       "      <td>2.77514</td>\n",
       "      <td>8.90767</td>\n",
       "      <td>1.47206</td>\n",
       "      <td>0.87271</td>\n",
       "      <td>4.22791</td>\n",
       "      <td>0.60317</td>\n",
       "      <td>0.46695</td>\n",
       "      <td>2.50889</td>\n",
       "      <td>4.48199</td>\n",
       "      <td>6.83279</td>\n",
       "      <td>1.09368</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                              ID  \\\n",
       "0  k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae|g__Bifidobacterium   \n",
       "1  k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Eggerthellales|f__Eggerthellaceae|g__Adlercreutzia           \n",
       "2  k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Eggerthellales|f__Eggerthellaceae|g__Asaccharobacter         \n",
       "3  k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Eggerthellales|f__Eggerthellaceae|g__Enterorhabdus           \n",
       "4  k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides                   \n",
       "\n",
       "   E15_S89_known_profiled_metagenome  E19b_S88_known_profiled_metagenome  \\\n",
       "0  0.00000                            33.99356                             \n",
       "1  0.01522                            0.02269                              \n",
       "2  0.06825                            0.02827                              \n",
       "3  1.68923                            1.13757                              \n",
       "4  2.57402                            1.03313                              \n",
       "\n",
       "   E28_S87_known_profiled_metagenome  E5b_S84_known_profiled_metagenome  \\\n",
       "0  0.00000                            0.00000                             \n",
       "1  0.00986                            0.01337                             \n",
       "2  0.02306                            0.02648                             \n",
       "3  0.90287                            1.69362                             \n",
       "4  2.58441                            5.52997                             \n",
       "\n",
       "   E7_S78_known_profiled_metagenome  E8_S74_known_profiled_metagenome  \\\n",
       "0  0.00000                           0.00000                            \n",
       "1  0.02927                           0.04169                            \n",
       "2  0.02673                           0.03658                            \n",
       "3  1.55661                           1.75809                            \n",
       "4  4.55413                           2.52983                            \n",
       "\n",
       "   R21b_S76_known_profiled_metagenome  R24_S82_known_profiled_metagenome  \\\n",
       "0  0.00000                             0.00000                             \n",
       "1  0.01290                             0.00000                             \n",
       "2  0.02223                             0.00543                             \n",
       "3  0.85820                             1.38410                             \n",
       "4  1.46428                             2.77514                             \n",
       "\n",
       "   R5_S75_known_profiled_metagenome  R6_S81_known_profiled_metagenome  \\\n",
       "0  0.00000                           0.00000                            \n",
       "1  0.02503                           0.05092                            \n",
       "2  0.03252                           0.06713                            \n",
       "3  1.53637                           2.20989                            \n",
       "4  8.90767                           1.47206                            \n",
       "\n",
       "   S11b_S73_known_profiled_metagenome  S11_S83_known_profiled_metagenome  \\\n",
       "0  0.00000                             0.00000                             \n",
       "1  0.01569                             0.03171                             \n",
       "2  0.02811                             0.04405                             \n",
       "3  1.07292                             2.22013                             \n",
       "4  0.87271                             4.22791                             \n",
       "\n",
       "   S13_S79_known_profiled_metagenome  S22b_S90_known_profiled_metagenome  \\\n",
       "0  0.00000                            20.50706                             \n",
       "1  0.04299                            0.01102                              \n",
       "2  0.03960                            0.01511                              \n",
       "3  1.54649                            0.47359                              \n",
       "4  0.60317                            0.46695                              \n",
       "\n",
       "   S22_S77_known_profiled_metagenome  S3b_S86_known_profiled_metagenome  \\\n",
       "0  0.00000                            11.21542                            \n",
       "1  0.02144                            0.01528                             \n",
       "2  0.00854                            0.01129                             \n",
       "3  0.81333                            0.61152                             \n",
       "4  2.50889                            4.48199                             \n",
       "\n",
       "   S4_S85_known_profiled_metagenome  S9_S80_known_profiled_metagenome  \n",
       "0  0.00000                           0.00000                           \n",
       "1  0.01784                           0.03676                           \n",
       "2  0.01096                           0.06209                           \n",
       "3  0.96265                           2.02247                           \n",
       "4  6.83279                           1.09368                           "
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here we have only the rows corresponding to genus level\n",
    "df_k_1_genus.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(38, 19)"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_k_1_genus.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merging the two pools on ID\n",
    "df_known_genus = pd.merge(df_k_1_genus,df_k_2_genus,on=['ID'],how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(38, 38)"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_known_genus.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean the ID column and sample name columns to create the abundance matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>E15_S89</th>\n",
       "      <th>E19b_S88</th>\n",
       "      <th>E28_S87</th>\n",
       "      <th>E5b_S84</th>\n",
       "      <th>E7_S78</th>\n",
       "      <th>E8_S74</th>\n",
       "      <th>R21b_S76</th>\n",
       "      <th>R24_S82</th>\n",
       "      <th>R5_S75</th>\n",
       "      <th>...</th>\n",
       "      <th>R14_S71</th>\n",
       "      <th>R21_S60</th>\n",
       "      <th>R26_S67</th>\n",
       "      <th>R29_S61</th>\n",
       "      <th>R2_S58</th>\n",
       "      <th>R6b_S54</th>\n",
       "      <th>R7b_S68</th>\n",
       "      <th>S20b_S59</th>\n",
       "      <th>S27_S72</th>\n",
       "      <th>S2b_S62</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae|g__Bifidobacterium</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>33.99356</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>70.28168</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>31.72024</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>10.61550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Eggerthellales|f__Eggerthellaceae|g__Adlercreutzia</td>\n",
       "      <td>0.01522</td>\n",
       "      <td>0.02269</td>\n",
       "      <td>0.00986</td>\n",
       "      <td>0.01337</td>\n",
       "      <td>0.02927</td>\n",
       "      <td>0.04169</td>\n",
       "      <td>0.0129</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.02503</td>\n",
       "      <td>...</td>\n",
       "      <td>0.04023</td>\n",
       "      <td>0.04791</td>\n",
       "      <td>0.01381</td>\n",
       "      <td>0.04743</td>\n",
       "      <td>0.04462</td>\n",
       "      <td>0.00712</td>\n",
       "      <td>0.02831</td>\n",
       "      <td>0.06023</td>\n",
       "      <td>0.01183</td>\n",
       "      <td>0.02226</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                              ID  \\\n",
       "0  k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae|g__Bifidobacterium   \n",
       "1  k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Eggerthellales|f__Eggerthellaceae|g__Adlercreutzia           \n",
       "\n",
       "   E15_S89  E19b_S88  E28_S87  E5b_S84   E7_S78   E8_S74  R21b_S76  R24_S82  \\\n",
       "0  0.00000  33.99356  0.00000  0.00000  0.00000  0.00000  0.0000    0.0       \n",
       "1  0.01522  0.02269   0.00986  0.01337  0.02927  0.04169  0.0129    0.0       \n",
       "\n",
       "    R5_S75  ...  R14_S71  R21_S60   R26_S67  R29_S61   R2_S58  R6b_S54  \\\n",
       "0  0.00000  ...  0.00000  0.00000  70.28168  0.00000  0.00000  0.00000   \n",
       "1  0.02503  ...  0.04023  0.04791  0.01381   0.04743  0.04462  0.00712   \n",
       "\n",
       "   R7b_S68  S20b_S59  S27_S72   S2b_S62  \n",
       "0  0.00000  31.72024  0.00000  10.61550  \n",
       "1  0.02831  0.06023   0.01183  0.02226   \n",
       "\n",
       "[2 rows x 38 columns]"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_known=clean_merged_table(df_known_genus,'_known_profiled_metagenome')\n",
    "df_known[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>E15_S89</th>\n",
       "      <th>E19b_S88</th>\n",
       "      <th>E28_S87</th>\n",
       "      <th>E5b_S84</th>\n",
       "      <th>E7_S78</th>\n",
       "      <th>E8_S74</th>\n",
       "      <th>R21b_S76</th>\n",
       "      <th>R24_S82</th>\n",
       "      <th>R5_S75</th>\n",
       "      <th>...</th>\n",
       "      <th>R14_S71</th>\n",
       "      <th>R21_S60</th>\n",
       "      <th>R26_S67</th>\n",
       "      <th>R29_S61</th>\n",
       "      <th>R2_S58</th>\n",
       "      <th>R6b_S54</th>\n",
       "      <th>R7b_S68</th>\n",
       "      <th>S20b_S59</th>\n",
       "      <th>S27_S72</th>\n",
       "      <th>S2b_S62</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Herpesvirales|f__Herpesviridae|g__Rhadinovirus</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.10895</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.11163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Ortervirales|f__Retroviridae|g__Betaretrovirus</td>\n",
       "      <td>0.23658</td>\n",
       "      <td>0.10169</td>\n",
       "      <td>0.09105</td>\n",
       "      <td>0.43274</td>\n",
       "      <td>0.28599</td>\n",
       "      <td>0.05585</td>\n",
       "      <td>0.03707</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.13411</td>\n",
       "      <td>...</td>\n",
       "      <td>0.06558</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.07879</td>\n",
       "      <td>0.22339</td>\n",
       "      <td>0.12315</td>\n",
       "      <td>0.04778</td>\n",
       "      <td>0.06023</td>\n",
       "      <td>0.02109</td>\n",
       "      <td>0.66410</td>\n",
       "      <td>1.23608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Ortervirales|f__Retroviridae|g__Gammaretrovirus</td>\n",
       "      <td>0.71291</td>\n",
       "      <td>0.06184</td>\n",
       "      <td>0.28809</td>\n",
       "      <td>1.54332</td>\n",
       "      <td>0.65427</td>\n",
       "      <td>0.37626</td>\n",
       "      <td>0.28503</td>\n",
       "      <td>0.06873</td>\n",
       "      <td>0.18652</td>\n",
       "      <td>...</td>\n",
       "      <td>0.13752</td>\n",
       "      <td>0.20025</td>\n",
       "      <td>0.19943</td>\n",
       "      <td>0.47122</td>\n",
       "      <td>0.38966</td>\n",
       "      <td>0.20041</td>\n",
       "      <td>0.88896</td>\n",
       "      <td>0.64872</td>\n",
       "      <td>1.68495</td>\n",
       "      <td>8.66917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Viruses_unclassified|f__Baculoviridae|g__Alphabaculovirus</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.03318</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Viruses_unclassified|f__Polydnaviridae|g__Bracovirus</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.08873</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.11210</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                         ID  \\\n",
       "33  k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Herpesvirales|f__Herpesviridae|g__Rhadinovirus              \n",
       "34  k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Ortervirales|f__Retroviridae|g__Betaretrovirus              \n",
       "35  k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Ortervirales|f__Retroviridae|g__Gammaretrovirus             \n",
       "36  k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Viruses_unclassified|f__Baculoviridae|g__Alphabaculovirus   \n",
       "37  k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Viruses_unclassified|f__Polydnaviridae|g__Bracovirus        \n",
       "\n",
       "    E15_S89  E19b_S88  E28_S87  E5b_S84   E7_S78   E8_S74  R21b_S76  R24_S82  \\\n",
       "33  0.00000  0.00000   0.10895  0.00000  0.00000  0.00000  0.00000   0.00000   \n",
       "34  0.23658  0.10169   0.09105  0.43274  0.28599  0.05585  0.03707   0.00000   \n",
       "35  0.71291  0.06184   0.28809  1.54332  0.65427  0.37626  0.28503   0.06873   \n",
       "36  0.00000  0.00000   0.00000  0.00000  0.00000  0.00000  0.00000   0.00000   \n",
       "37  0.00000  0.00000   0.00000  0.00000  0.00000  0.00000  0.00000   0.00000   \n",
       "\n",
       "     R5_S75  ...  R14_S71  R21_S60  R26_S67  R29_S61   R2_S58  R6b_S54  \\\n",
       "33  0.00000  ...  0.00000  0.00000  0.00000  0.00000  0.00000  0.00000   \n",
       "34  0.13411  ...  0.06558  0.00000  0.07879  0.22339  0.12315  0.04778   \n",
       "35  0.18652  ...  0.13752  0.20025  0.19943  0.47122  0.38966  0.20041   \n",
       "36  0.03318  ... NaN      NaN      NaN      NaN      NaN      NaN        \n",
       "37  0.00000  ...  0.00000  0.08873  0.00000  0.00000  0.00000  0.00000   \n",
       "\n",
       "    R7b_S68  S20b_S59  S27_S72  S2b_S62  \n",
       "33  0.00000  0.00000   0.00000  0.11163  \n",
       "34  0.06023  0.02109   0.66410  1.23608  \n",
       "35  0.88896  0.64872   1.68495  8.66917  \n",
       "36 NaN      NaN       NaN      NaN       \n",
       "37  0.00000  0.00000   0.00000  0.11210  \n",
       "\n",
       "[5 rows x 38 columns]"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_known.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace NaN with 0\n",
    "df_known.fillna(0,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_column = df_known[['ID']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E15_S89</th>\n",
       "      <th>E19b_S88</th>\n",
       "      <th>E28_S87</th>\n",
       "      <th>E5b_S84</th>\n",
       "      <th>E7_S78</th>\n",
       "      <th>E8_S74</th>\n",
       "      <th>R21b_S76</th>\n",
       "      <th>R24_S82</th>\n",
       "      <th>R5_S75</th>\n",
       "      <th>R6_S81</th>\n",
       "      <th>...</th>\n",
       "      <th>R21_S60</th>\n",
       "      <th>R26_S67</th>\n",
       "      <th>R29_S61</th>\n",
       "      <th>R2_S58</th>\n",
       "      <th>R6b_S54</th>\n",
       "      <th>R7b_S68</th>\n",
       "      <th>S20b_S59</th>\n",
       "      <th>S27_S72</th>\n",
       "      <th>S2b_S62</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>33.99356</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>70.28168</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>31.72024</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>10.61550</td>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae|g__Bifidobacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.44706</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Eggerthellales|f__Eggerthellaceae|g__Enterorhabdus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>5.52997</td>\n",
       "      <td>4.55413</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>8.90767</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.36496</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>3.20713</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>32.67238</td>\n",
       "      <td>15.48173</td>\n",
       "      <td>19.72849</td>\n",
       "      <td>29.41141</td>\n",
       "      <td>31.89512</td>\n",
       "      <td>37.51055</td>\n",
       "      <td>5.50114</td>\n",
       "      <td>12.37102</td>\n",
       "      <td>38.94397</td>\n",
       "      <td>47.76264</td>\n",
       "      <td>...</td>\n",
       "      <td>16.31096</td>\n",
       "      <td>6.75982</td>\n",
       "      <td>32.03713</td>\n",
       "      <td>14.61179</td>\n",
       "      <td>10.52932</td>\n",
       "      <td>15.37850</td>\n",
       "      <td>17.26982</td>\n",
       "      <td>22.22244</td>\n",
       "      <td>26.37850</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Muribaculaceae|g__Muribaculaceae_unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>25.68020</td>\n",
       "      <td>9.66215</td>\n",
       "      <td>15.47967</td>\n",
       "      <td>25.57575</td>\n",
       "      <td>26.25145</td>\n",
       "      <td>14.94594</td>\n",
       "      <td>10.05876</td>\n",
       "      <td>6.31194</td>\n",
       "      <td>19.81598</td>\n",
       "      <td>7.73166</td>\n",
       "      <td>...</td>\n",
       "      <td>31.06341</td>\n",
       "      <td>8.38871</td>\n",
       "      <td>29.84066</td>\n",
       "      <td>10.03842</td>\n",
       "      <td>9.99244</td>\n",
       "      <td>11.13738</td>\n",
       "      <td>6.07923</td>\n",
       "      <td>39.77292</td>\n",
       "      <td>10.81153</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Muribaculaceae|g__Muribaculum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4.98770</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>6.95289</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>22.89472</td>\n",
       "      <td>3.14556</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.29901</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>3.62847</td>\n",
       "      <td>20.39355</td>\n",
       "      <td>23.80294</td>\n",
       "      <td>15.56993</td>\n",
       "      <td>4.02813</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Prevotellaceae|g__Prevotella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>11.98444</td>\n",
       "      <td>6.50151</td>\n",
       "      <td>7.27066</td>\n",
       "      <td>4.77874</td>\n",
       "      <td>7.06418</td>\n",
       "      <td>6.39057</td>\n",
       "      <td>7.94512</td>\n",
       "      <td>6.45230</td>\n",
       "      <td>11.26780</td>\n",
       "      <td>15.46397</td>\n",
       "      <td>...</td>\n",
       "      <td>17.59737</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>7.57646</td>\n",
       "      <td>5.05002</td>\n",
       "      <td>3.38600</td>\n",
       "      <td>3.54435</td>\n",
       "      <td>5.69944</td>\n",
       "      <td>3.89995</td>\n",
       "      <td>3.61393</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Flavobacteriia|o__Flavobacteriales|f__Flavobacteriaceae|g__Imtechella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>14.32023</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>10.29383</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>k__Bacteria|p__Deferribacteres|c__Deferribacteres|o__Deferribacterales|f__Deferribacteraceae|g__Mucispirillum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6.56370</td>\n",
       "      <td>3.93777</td>\n",
       "      <td>12.25170</td>\n",
       "      <td>3.03387</td>\n",
       "      <td>13.31346</td>\n",
       "      <td>6.56068</td>\n",
       "      <td>16.78792</td>\n",
       "      <td>40.71146</td>\n",
       "      <td>11.46063</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>15.52980</td>\n",
       "      <td>3.70173</td>\n",
       "      <td>10.77961</td>\n",
       "      <td>7.31977</td>\n",
       "      <td>20.50488</td>\n",
       "      <td>32.77265</td>\n",
       "      <td>5.87253</td>\n",
       "      <td>8.71473</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Eubacteriaceae|g__Eubacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>11.74744</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>12.64291</td>\n",
       "      <td>7.55550</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>3.03219</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.54163</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.00844</td>\n",
       "      <td>5.65651</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.63524</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>3.83587</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Lachnospiraceae|g__Lachnospiraceae_unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>5.80794</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.54464</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Anaerotruncus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>7.98535</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>19.38527</td>\n",
       "      <td>3.47151</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>20.05666</td>\n",
       "      <td>4.84151</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>6.06839</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>9.09260</td>\n",
       "      <td>42.98675</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Erysipelotrichia|o__Erysipelotrichales|f__Erysipelotrichaceae|g__Faecalibaculum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>13.52005</td>\n",
       "      <td>7.97246</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>3.24939</td>\n",
       "      <td>27.48096</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>12.92017</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.86170</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.12568</td>\n",
       "      <td>10.68457</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.42385</td>\n",
       "      <td>5.22013</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales|f__Akkermansiaceae|g__Akkermansia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>3.89266</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>6.71008</td>\n",
       "      <td>19.94497</td>\n",
       "      <td>k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Herpesvirales|f__Alloherpesviridae|g__Cyprinivirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>8.66917</td>\n",
       "      <td>k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Ortervirales|f__Retroviridae|g__Gammaretrovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>10.12621</td>\n",
       "      <td>5.15580</td>\n",
       "      <td>10.95882</td>\n",
       "      <td>11.66320</td>\n",
       "      <td>9.36616</td>\n",
       "      <td>11.28621</td>\n",
       "      <td>4.48988</td>\n",
       "      <td>7.84736</td>\n",
       "      <td>9.60395</td>\n",
       "      <td>11.57992</td>\n",
       "      <td>...</td>\n",
       "      <td>6.68399</td>\n",
       "      <td>6.00636</td>\n",
       "      <td>10.67355</td>\n",
       "      <td>8.23067</td>\n",
       "      <td>8.55888</td>\n",
       "      <td>8.99923</td>\n",
       "      <td>8.72971</td>\n",
       "      <td>9.43161</td>\n",
       "      <td>8.37877</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     E15_S89  E19b_S88   E28_S87   E5b_S84    E7_S78    E8_S74  R21b_S76  \\\n",
       "0   0.00000   33.99356  0.00000   0.00000   0.00000   0.00000   0.00000    \n",
       "1   0.00000   0.00000   0.00000   0.00000   0.00000   0.00000   0.00000    \n",
       "2   0.00000   0.00000   0.00000   5.52997   4.55413   0.00000   0.00000    \n",
       "3   32.67238  15.48173  19.72849  29.41141  31.89512  37.51055  5.50114    \n",
       "4   25.68020  9.66215   15.47967  25.57575  26.25145  14.94594  10.05876   \n",
       "5   4.98770   0.00000   6.95289   0.00000   0.00000   0.00000   22.89472   \n",
       "6   11.98444  6.50151   7.27066   4.77874   7.06418   6.39057   7.94512    \n",
       "7   0.00000   0.00000   0.00000   0.00000   0.00000   0.00000   0.00000    \n",
       "8   6.56370   3.93777   12.25170  3.03387   13.31346  6.56068   16.78792   \n",
       "9   0.00000   0.00000   0.00000   0.00000   0.00000   0.00000   0.00000    \n",
       "10  0.00000   11.74744  0.00000   12.64291  7.55550   0.00000   0.00000    \n",
       "11  0.00000   0.00000   0.00000   0.00000   0.00000   0.00000   0.00000    \n",
       "12  7.98535   0.00000   19.38527  3.47151   0.00000   20.05666  4.84151    \n",
       "13  0.00000   13.52005  7.97246   0.00000   0.00000   3.24939   27.48096   \n",
       "14  0.00000   0.00000   0.00000   3.89266   0.00000   0.00000   0.00000    \n",
       "15  0.00000   0.00000   0.00000   0.00000   0.00000   0.00000   0.00000    \n",
       "16  10.12621  5.15580   10.95882  11.66320  9.36616   11.28621  4.48988    \n",
       "\n",
       "     R24_S82    R5_S75    R6_S81  ...   R21_S60   R26_S67   R29_S61    R2_S58  \\\n",
       "0   0.00000   0.00000   0.00000   ...  0.00000   70.28168  0.00000   0.00000    \n",
       "1   0.00000   0.00000   0.00000   ...  3.44706   0.00000   0.00000   0.00000    \n",
       "2   0.00000   8.90767   0.00000   ...  0.00000   0.00000   0.00000   0.00000    \n",
       "3   12.37102  38.94397  47.76264  ...  16.31096  6.75982   32.03713  14.61179   \n",
       "4   6.31194   19.81598  7.73166   ...  31.06341  8.38871   29.84066  10.03842   \n",
       "5   3.14556   0.00000   0.00000   ...  3.29901   0.00000   0.00000   3.62847    \n",
       "6   6.45230   11.26780  15.46397  ...  17.59737  0.00000   7.57646   5.05002    \n",
       "7   14.32023  0.00000   0.00000   ...  0.00000   0.00000   0.00000   0.00000    \n",
       "8   40.71146  11.46063  0.00000   ...  15.52980  3.70173   10.77961  7.31977    \n",
       "9   0.00000   0.00000   0.00000   ...  0.00000   0.00000   0.00000   0.00000    \n",
       "10  3.03219   0.00000   4.54163   ...  0.00000   0.00000   0.00000   4.00844    \n",
       "11  5.80794   0.00000   0.00000   ...  0.00000   0.00000   0.00000   0.00000    \n",
       "12  0.00000   0.00000   0.00000   ...  6.06839   0.00000   9.09260   42.98675   \n",
       "13  0.00000   0.00000   12.92017  ...  0.00000   4.86170   0.00000   4.12568    \n",
       "14  0.00000   0.00000   0.00000   ...  0.00000   0.00000   0.00000   0.00000    \n",
       "15  0.00000   0.00000   0.00000   ...  0.00000   0.00000   0.00000   0.00000    \n",
       "16  7.84736   9.60395   11.57992  ...  6.68399   6.00636   10.67355  8.23067    \n",
       "\n",
       "     R6b_S54   R7b_S68  S20b_S59   S27_S72   S2b_S62  \\\n",
       "0   0.00000   0.00000   31.72024  0.00000   10.61550   \n",
       "1   0.00000   0.00000   0.00000   0.00000   0.00000    \n",
       "2   0.00000   4.36496   0.00000   0.00000   3.20713    \n",
       "3   10.52932  15.37850  17.26982  22.22244  26.37850   \n",
       "4   9.99244   11.13738  6.07923   39.77292  10.81153   \n",
       "5   20.39355  23.80294  15.56993  4.02813   0.00000    \n",
       "6   3.38600   3.54435   5.69944   3.89995   3.61393    \n",
       "7   10.29383  0.00000   0.00000   0.00000   0.00000    \n",
       "8   20.50488  32.77265  5.87253   8.71473   0.00000    \n",
       "9   0.00000   0.00000   0.00000   0.00000   0.00000    \n",
       "10  5.65651   0.00000   4.63524   0.00000   3.83587    \n",
       "11  0.00000   0.00000   0.00000   0.00000   4.54464    \n",
       "12  0.00000   0.00000   0.00000   0.00000   0.00000    \n",
       "13  10.68457  0.00000   4.42385   5.22013   0.00000    \n",
       "14  0.00000   0.00000   0.00000   6.71008   19.94497   \n",
       "15  0.00000   0.00000   0.00000   0.00000   8.66917    \n",
       "16  8.55888   8.99923   8.72971   9.43161   8.37877    \n",
       "\n",
       "                                                                                                                  ID  \n",
       "0   k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae|g__Bifidobacterium     \n",
       "1   k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Eggerthellales|f__Eggerthellaceae|g__Enterorhabdus             \n",
       "2   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides                     \n",
       "3   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Muribaculaceae|g__Muribaculaceae_unclassified     \n",
       "4   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Muribaculaceae|g__Muribaculum                     \n",
       "5   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Prevotellaceae|g__Prevotella                      \n",
       "6   k__Bacteria|p__Bacteroidetes|c__Flavobacteriia|o__Flavobacteriales|f__Flavobacteriaceae|g__Imtechella             \n",
       "7   k__Bacteria|p__Deferribacteres|c__Deferribacteres|o__Deferribacterales|f__Deferribacteraceae|g__Mucispirillum     \n",
       "8   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus                      \n",
       "9   k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Eubacteriaceae|g__Eubacterium                         \n",
       "10  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Lachnospiraceae|g__Lachnospiraceae_unclassified       \n",
       "11  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Anaerotruncus                      \n",
       "12  k__Bacteria|p__Firmicutes|c__Erysipelotrichia|o__Erysipelotrichales|f__Erysipelotrichaceae|g__Faecalibaculum      \n",
       "13  k__Bacteria|p__Verrucomicrobia|c__Verrucomicrobiae|o__Verrucomicrobiales|f__Akkermansiaceae|g__Akkermansia        \n",
       "14  k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Herpesvirales|f__Alloherpesviridae|g__Cyprinivirus  \n",
       "15  k__Viruses|p__Viruses_unclassified|c__Viruses_unclassified|o__Ortervirales|f__Retroviridae|g__Gammaretrovirus     \n",
       "16  Other                                                                                                             \n",
       "\n",
       "[17 rows x 38 columns]"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abund_matrix=create_other_group(df_known,3) # Setting 3 as threshold - anything taxa comprising 3% or less of the matrix will be grouped into other\n",
    "abund_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "abun_matrix=add_otu_primary_key(df_known)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>E15_S89</th>\n",
       "      <th>E19b_S88</th>\n",
       "      <th>E28_S87</th>\n",
       "      <th>E5b_S84</th>\n",
       "      <th>E7_S78</th>\n",
       "      <th>E8_S74</th>\n",
       "      <th>R21b_S76</th>\n",
       "      <th>R24_S82</th>\n",
       "      <th>R5_S75</th>\n",
       "      <th>...</th>\n",
       "      <th>R21_S60</th>\n",
       "      <th>R26_S67</th>\n",
       "      <th>R29_S61</th>\n",
       "      <th>R2_S58</th>\n",
       "      <th>R6b_S54</th>\n",
       "      <th>R7b_S68</th>\n",
       "      <th>S20b_S59</th>\n",
       "      <th>S27_S72</th>\n",
       "      <th>S2b_S62</th>\n",
       "      <th>Otu</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae|g__Bifidobacterium</td>\n",
       "      <td>0.0</td>\n",
       "      <td>33.99356</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>70.28168</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.72024</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.6155</td>\n",
       "      <td>Otu0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                              ID  \\\n",
       "0  k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Bifidobacteriales|f__Bifidobacteriaceae|g__Bifidobacterium   \n",
       "\n",
       "   E15_S89  E19b_S88  E28_S87  E5b_S84  E7_S78  E8_S74  R21b_S76  R24_S82  \\\n",
       "0  0.0      33.99356  0.0      0.0      0.0     0.0     0.0       0.0       \n",
       "\n",
       "   R5_S75  ...  R21_S60   R26_S67  R29_S61  R2_S58  R6b_S54  R7b_S68  \\\n",
       "0  0.0     ...  0.0      70.28168  0.0      0.0     0.0      0.0       \n",
       "\n",
       "   S20b_S59  S27_S72  S2b_S62   Otu  \n",
       "0  31.72024  0.0      10.6155  Otu0  \n",
       "\n",
       "[1 rows x 39 columns]"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abun_matrix[0:1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create taxa table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Otu</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Viruses</td>\n",
       "      <td>Viruses_unclassified</td>\n",
       "      <td>Viruses_unclassified</td>\n",
       "      <td>Ortervirales</td>\n",
       "      <td>Retroviridae</td>\n",
       "      <td>Betaretrovirus</td>\n",
       "      <td>Otu34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Viruses</td>\n",
       "      <td>Viruses_unclassified</td>\n",
       "      <td>Viruses_unclassified</td>\n",
       "      <td>Ortervirales</td>\n",
       "      <td>Retroviridae</td>\n",
       "      <td>Gammaretrovirus</td>\n",
       "      <td>Otu35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Viruses</td>\n",
       "      <td>Viruses_unclassified</td>\n",
       "      <td>Viruses_unclassified</td>\n",
       "      <td>Viruses_unclassified</td>\n",
       "      <td>Baculoviridae</td>\n",
       "      <td>Alphabaculovirus</td>\n",
       "      <td>Otu36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Viruses</td>\n",
       "      <td>Viruses_unclassified</td>\n",
       "      <td>Viruses_unclassified</td>\n",
       "      <td>Viruses_unclassified</td>\n",
       "      <td>Polydnaviridae</td>\n",
       "      <td>Bracovirus</td>\n",
       "      <td>Otu37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Other</td>\n",
       "      <td>Other</td>\n",
       "      <td>Other</td>\n",
       "      <td>Other</td>\n",
       "      <td>Other</td>\n",
       "      <td>Other</td>\n",
       "      <td>Otu38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Kingdom                Phylum                 Class                 Order  \\\n",
       "34  Viruses  Viruses_unclassified  Viruses_unclassified  Ortervirales           \n",
       "35  Viruses  Viruses_unclassified  Viruses_unclassified  Ortervirales           \n",
       "36  Viruses  Viruses_unclassified  Viruses_unclassified  Viruses_unclassified   \n",
       "37  Viruses  Viruses_unclassified  Viruses_unclassified  Viruses_unclassified   \n",
       "38  Other    Other                 Other                 Other                  \n",
       "\n",
       "            Family             Genus    Otu  \n",
       "34  Retroviridae    Betaretrovirus    Otu34  \n",
       "35  Retroviridae    Gammaretrovirus   Otu35  \n",
       "36  Baculoviridae   Alphabaculovirus  Otu36  \n",
       "37  Polydnaviridae  Bracovirus        Otu37  \n",
       "38  Other           Other             Otu38  "
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genus_taxa=get_taxa_columns(abun_matrix,'Genus')\n",
    "genus_taxa.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create sample data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['E15_S89', 'E19b_S88', 'E28_S87', 'E5b_S84', 'E7_S78', 'E8_S74', 'R21b_S76']"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extracts all columns containing a number in name\n",
    "sample_cols=get_sample_cols(abun_matrix)\n",
    "sample_cols[0:7]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Behaviour</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E15_S89</td>\n",
       "      <td>Extinction</td>\n",
       "      <td>murine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E19b_S88</td>\n",
       "      <td>Extinction</td>\n",
       "      <td>murine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E28_S87</td>\n",
       "      <td>Extinction</td>\n",
       "      <td>murine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E5b_S84</td>\n",
       "      <td>Extinction</td>\n",
       "      <td>murine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E7_S78</td>\n",
       "      <td>Extinction</td>\n",
       "      <td>murine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample   Behaviour    Type\n",
       "0  E15_S89   Extinction  murine\n",
       "1  E19b_S88  Extinction  murine\n",
       "2  E28_S87   Extinction  murine\n",
       "3  E5b_S84   Extinction  murine\n",
       "4  E7_S78    Extinction  murine"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df=create_sample_df(abund_matrix)\n",
    "sample_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export to csv for importing into phyloseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "abund_matrix.to_csv('data/genus_abundance.csv',index=False)\n",
    "genus_taxa.to_csv('data/genus_taxa.csv',index=False)\n",
    "sample_df.to_csv('data/sample_data.csv',index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
